codec/src/main/java/io/netty/handler/codec/BatchFrameDecoder.java

```diff
@@ -0,0 +1,114 @@
+/*
+ * Copyright 2016 The Netty Project
+ *
+ * The Netty Project licenses this file to you under the Apache License,
+ * version 2.0 (the "License"); you may not use this file except in compliance
+ * with the License. You may obtain a copy of the License at:
+ *
+ *   http://www.apache.org/licenses/LICENSE-2.0
+ *
+ * Unless required by applicable law or agreed to in writing, software
+ * distributed under the License is distributed on an "AS IS" BASIS, WITHOUT
+ * WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied. See the
+ * License for the specific language governing permissions and limitations
+ * under the License.
+ */
+
+package io.netty.handler.codec;
+
+import io.netty.buffer.ByteBuf;
+import io.netty.channel.ChannelHandlerContext;
+import io.netty.channel.ChannelInboundHandlerAdapter;
+
+import java.util.List;
+
+public abstract class BatchFrameDecoder extends ChannelInboundHandlerAdapter {
+
+    private final int startBufferSize;
+
+    private ByteBuf buffer;
+
+    /**
+     * @param startBufferSize the start buffer size which is used for batching incoming {@link ByteBuf} messages. Size
+     * of this buffer must be greater or equal to zero. If the size of this buffer is zero then the default size will be
+     * used (= 256 bytes).
+     */
+    protected BatchFrameDecoder(int startBufferSize) {
+        if (startBufferSize < 0) {
+            throw new IllegalArgumentException("startBufferSize must be greater or equal to zero: " + startBufferSize);
+        }
+
+        this.startBufferSize = startBufferSize;
+
+        CodecUtil.ensureNotSharable(this);
+    }
+
+    protected abstract void decode(ChannelHandlerContext ctx, ByteBuf buffer, List<Object> out);
+
+    @Override
+    public void channelActive(ChannelHandlerContext ctx) throws Exception {
+        buffer = internalBuffer(ctx, startBufferSize);
+
+        ctx.fireChannelActive();
+    }
+
+    @Override
+    public void channelInactive(ChannelHandlerContext ctx) throws Exception {
+        releaseInternalBuffer();
+        ctx.fireChannelInactive();
+    }
+
+    @Override
+    public void handlerRemoved(ChannelHandlerContext ctx) throws Exception {
+        releaseInternalBuffer();
+    }
+
+    @Override
+    public void channelRead(ChannelHandlerContext ctx, Object msg) throws Exception {
+        if (msg instanceof ByteBuf) {
+            final ByteBuf frame = (ByteBuf) msg;
+
+            try {
+                buffer.writeBytes(frame);
```

Comment:
So, I looked at `ByteToMessageDecoder` more closely and I think it wouldn't be easy to modify it so that we could reuse some code and change its behavior (make the `decode` method invocation optional in `channelReadComplete`). Let's start with extracting the `Cumulator` classes and reusing it in the new batch frame decoders. I think it would be much easier. WDYT?